sakura_core/Makefile

```
@@ -55,424 +84,61 @@ LIBS= \
-util/window.o \
-view/CCaret.o \
-view/CEditView.o \
-view/CEditView_Cmdgrep.o \
-view/CEditView_CmdHokan.o \
-view/CEditView_Cmdisrch.o \
-view/CEditView_Command.o \
-view/CEditView_Command_New.o \
-view/CEditView_Diff.o \
-view/CEditView_ExecCmd.o \
-view/CEditView_Ime.o \
-view/CEditView_Mouse.o \
-view/CEditView_Paint.o \
-view/CEditView_Paint_Bracket.o \
-view/CEditView_Scroll.o \
-view/CEditView_Search.o \
-view/CRuler.o \
-view/CTextArea.o \
-view/CTextDrawer.o \
-view/CTextMetrics.o \
-view/CViewCalc.o \
-view/CViewFont.o \
-view/CViewParser.o \
-view/CViewSelect.o \
-view/DispPos.o \
-view/colors/CColorStrategy.o \
-view/colors/CColor_Comment.o \
-view/colors/CColor_Found.o \
-view/colors/CColor_Heredoc.o \
-view/colors/CColor_KeywordSet.o \
-view/colors/CColor_Numeric.o \
-view/colors/CColor_Quote.o \
-view/colors/CColor_RegexKeyword.o \
-view/colors/CColor_Url.o \
-view/figures/CFigureManager.o \
-view/figures/CFigureStrategy.o \
-view/figures/CFigure_Comma.o \
-view/figures/CFigure_CtrlCode.o \
-view/figures/CFigure_Eol.o \
-view/figures/CFigure_HanSpace.o \
-view/figures/CFigure_Tab.o \
-view/figures/CFigure_ZenSpace.o \
-window/CAutoScrollWnd.o \
-window/CEditWnd.o \
-window/CMainStatusBar.o \
-window/CMainToolBar.o \
-window/CSplitBoxWnd.o \
-window/CSplitterWnd.o \
-window/CTabWnd.o \
-window/CTipWnd.o \
-window/CWnd.o \
-_main/CAppMode.o \
-_main/CCommandLine.o \
-_main/CControlProcess.o \
-_main/CControlTray.o \
-_main/CNormalProcess.o \
-_main/CProcess.o \
-_main/CProcessFactory.o \
-_main/global.o \
-_main/WinMain.o \
-_os/CClipboard.o \
-_os/CDropTarget.o \
-sakura_rc.o \
+SRCS = $(wildcard $(SRCDIR)/*.cpp) \
+       $(wildcard $(SRCDIR)/*/*.cpp) \
+       $(wildcard $(SRCDIR)/*/*/*.cpp)
+OBJS = $(SRCS:$(SRCDIR)/%.cpp=%.o) sakura_rc.o
+
+_DIRS = $(filter %/, $(wildcard $(SRCDIR)/*/)) \
+        $(filter %/, $(wildcard $(SRCDIR)/*/*/))
+DIRS = $(_DIRS:$(SRCDIR)/%/=%)
 
 DEPS= $(OBJS:%.o=%.d) StdAfx.h.d
 
 GENERATED_FILES= \
-Funccode_define.h \
-Funccode_enum.h \
-githash.h \
+ Funccode_define.h \
+ Funccode_enum.h \
+ githash.h \
 
-HEADERMAKETOOLDIR= ../HeaderMake
-HEADERMAKE= $(HEADERMAKETOOLDIR)/HeaderMake.exe
+HEADERMAKETOOLDIR= $(SRCDIR)/../HeaderMake
+HEADERMAKE= $(or $(OUTDIR),$(HEADERMAKETOOLDIR))/HeaderMake.exe
 
 all: $(exe)
 
 $(exe): $(OBJS)
 	$(CXX) -o $@ $(OBJS) $(LIBS)
 
-Funccode_define.h: $(HEADERMAKE) Funccode_x.hsrc
-	$(HEADERMAKE) -in=../sakura_core/Funccode_x.hsrc -out=../sakura_core/Funccode_define.h -mode=define
+Funccode_define.h: Funccode_x.hsrc $(HEADERMAKE)
+	$(HEADERMAKE) -in=$< -out=$@ -mode=define
 
-Funccode_enum.h: $(HEADERMAKE) Funccode_x.hsrc
-	$(HEADERMAKE) -in=../sakura_core/Funccode_x.hsrc -out=../sakura_core/Funccode_enum.h -mode=enum -enum=EFunctionCode
+Funccode_enum.h: Funccode_x.hsrc $(HEADERMAKE)
+	$(HEADERMAKE) -in=$< -out=$@ -mode=enum -enum=EFunctionCode
 
 githash.h:
-	cmd /c ..\sakura\githash.bat ../sakura_core
+	cmd /c $(subst /,\,$(SRCDIR))\..\sakura\githash.bat .
```

ローカルで試した感じ、ここがうまく行きませんでした。

```suggestion
	cmd /c $(realpath $(SRCDIR)/../sakura/githash.bat) .
```

相対パスの解決がうまく行かなかった感じ・・・

---

私のところでは問題なくビルドできているので、上手くいかなかった手順を教えてください。

---

ビルドできるようになってますね。

1. 85288f1f895ec1b29bb084d266f5a8484cf70499 をチェックアウト
1. mkdir build
1. cd build
1. mingw32-make -f ../sakura_core/Makefile OUTDIR=.

---

あれ？それでうまく行くのであれば、`$(SRCDIR))\..\sakura\githash.bat`ではなく`$(SRCDIR))/../sakura/githash.bat`と書けばsubstする必要なくないですか？

---

ユーザーが `SRCDIR` を指定した場合、かつ、パス区切りを `\\` で書いたときの対策、ということで理解しました。
とするとユーザーに `SRCDIR` を指定させないほうが楽な気がしてきました。

---

> ビルドできるようになってますね。

というか私は mintty 上をメインで作業してたので、基本的にはずっとビルドできてたはずなのですがｗ

> `$(SRCDIR))/../sakura/githash.bat`と書けばsubstする必要なくないですか？

cmd.exe は `/` 区切りのパスを認識してくれないので、`/` を `\` に変える必要があります。